pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -237,78 +241,183 @@ func (s *GatewayClient) QueryIndex(_ context.Context, _ *logproto.QueryIndexRequ
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.GetVolume(ctx, in, opts...)
+		resp, err = client.GetVolume(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
+func (s *GatewayClient) GetShards(
+	ctx context.Context,
+	in *logproto.ShardsRequest,
+) (res *logproto.ShardsResponse, err error) {
+
+	// We try to get the shards from the index gateway,
+	// but if it's not implemented, we fall back to the stats.
+	// We limit the maximum number of errors to 2 to avoid
+	// cascading all requests to new node(s) when
+	// the idx-gw replicas start to update to a version
+	// which supports the new API.
+	var (
+		maxErrs = 2
+		errCt   int
+	)
+
+	if err := s.poolDoWithStrategy(
+		ctx,
+		func(client logproto.IndexGatewayClient) error {
+			perReplicaResult := &logproto.ShardsResponse{}
+			streamer, err := client.GetShards(ctx, in)
+			if err != nil {
+				return errors.Wrap(err, "get shards")
+			}
+
+			// TODO(owen-d): stream currently unused (buffered) because query planning doesn't expect a streamed response,
+			// but can be improved easily in the future by using a stream here.
+			for {
+				resp, err := streamer.Recv()
+				if err == io.EOF {
+					break
+				}
+				if err != nil {
+					return errors.WithStack(err)
+				}
+				perReplicaResult.Shards = append(perReplicaResult.Shards, resp.Shards...)
+			}
+
+			// Since `poolDo` retries on error, we only want to set the response if we got a successful response.
+			// This avoids cases where we add duplicates to the response on retries.
+			res = perReplicaResult
+
+			return nil
+		},
+		func(err error) bool {
+			errCt++
+			return errCt <= maxErrs
+		},
+	); err != nil {
+		if isUnimplementedCallError(err) {
+			return s.getShardsFromStatsFallback(ctx, in)
+		}
+		return nil, err
+	}
+	return res, nil
+}
+
+func (s *GatewayClient) getShardsFromStatsFallback(
+	ctx context.Context,
+	in *logproto.ShardsRequest,
+) (*logproto.ShardsResponse, error) {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "index gateway client get tenant ID")
+	}
+
+	p, err := indexgateway.ExtractShardRequestMatchersAndAST(in.Query)
+	if err != nil {
+		return nil, errors.Wrap(err, "failure while falling back to stats for shard calculation")
+
+	}
+
+	stats, err := s.GetStats(
+		ctx,
+		&logproto.IndexStatsRequest{
+			From:     in.From,
+			Through:  in.Through,
+			Matchers: (&syntax.MatchersExpr{Mts: p.Matchers}).String(),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var strategy sharding.PowerOfTwoSharding
+	shards := strategy.ShardsFor(stats.Bytes, uint64(s.limits.TSDBMaxBytesPerShard(userID)))
+	return &logproto.ShardsResponse{
+		Shards: shards,
+	}, nil
+}
+
+// TODO(owen-d): this was copied from ingester_querier.go -- move it to a shared pkg
+// isUnimplementedCallError tells if the GRPC error is a gRPC error with code Unimplemented.
+func isUnimplementedCallError(err error) bool {
```

Comment:
Maybe we can extract this to the common `util` pkg, inside the [errors.go file.](https://github.com/grafana/loki/blob/08e505dd50915d2bdbb678211449e56d6c84ac08/pkg/util/errors.go#L150-L149)